File: DSA/expensive_expeditions/part4_medium/javascript/judge.js
const Solution = require('./solution');

class Judge {
    runTests() {
        const testCasesM = [
            [
                [1, -1, 1, 1, 1, 1],
                [3, 0, -4, 4, 4, 2],
                [1, 0, 1, 1, 1, 1],
                [1, 0, 0, 0, 0, -2],
                [0, 1, 0, 1, -1, 1],
                [1, 2, -4, 1, 1, 1]
            ],
            [
                [1, -1, 1, 1, 1, 1],
                [3, 0, -4, 4, 4, 2],
                [1, 0, 1, 1, 1, 1],
                [1, 0, 0, 0, 0, -2],
                [0, 1, 0, 1, -1, 1],
                [1, 2, -4, 1, 1, 1]
            ]
        ];

        const testCasesN = [
            5,
            3
        ];

        for(let i = 0; i < testCasesM.length; i++) {
            this.test(testCasesM[i], testCasesN[i], this.testCaseSolutionOutput(testCasesM[i], testCasesN[i]));
        }
    }

    test(m, n, expected) {
        const solution = new Solution();
        const result = solution.findMostValuableArea(m, n);
        const passed = result === expected;
        console.log(`Matrix: ${m}`);
        console.log(`n: ${n}`);
        console.log(`Expected: ${expected}`);
        console.log(`Your Output: ${result}`);
        console.log(passed ? "Test PASSED" : "Test FAILED");
        console.log("-----------");
    }

    testCaseSolutionOutput(m, n) {
        const rows = m.length;
        
        if (rows === 0 || n > rows || n > m[0].length) return [[-1, -1], [-1, -1]]; // Test cases should always be valid, so we return [-1, -1] if not
        const cols = m[0].length;

        const prefixSum = new Array(rows + 1).fill(0).map(() => new Array(cols + 1).fill(0));
        for(let i = 1; i <= rows; i++) {
            for(let j = 1; j <= cols; j++) {
                prefixSum[i][j] = m[i - 1][j - 1] + prefixSum[i - 1][j] + prefixSum[i][j - 1] - prefixSum[i - 1][j - 1];
            }
        }

        let maxSum = -Infinity;
        let topLeft = [0, 0];
        let bottomRight = [0, 0];

        for(let i = n; i <= rows; i++) {
            for(let j = n; j <= cols; j++) {
                const currentSum = prefixSum[i][j] - prefixSum[i - n][j] - prefixSum[i][j - n] + prefixSum[i - n][j - n];
                if(currentSum > maxSum) {
                    maxSum = currentSum;
                    topLeft = [i - n, j - n];
                    bottomRight = [i - 1, j - 1];
                }
            }
        }

        return [topLeft, bottomRight];
    }
}

module.exports = Judge;

File: DSA/troubling_traversals/part4_hard/java/Judge.java
package DSA.troubling_traversals.part4_hard.java;

import java.util.*;

public class Judge {

    public void runTests() {
        int[][][] testCasesF = {
            {
                {1, 1, 1, 1, 1, 1},
                {3, 0, 4, 4, 4, 2},
                {1, 0, 1, 1, 1, 1},
                {1, 0, 0, 0, 0, 2}, 
                {0, 1, 0, 1, 1, 1},
                {1, 2, 4, 1, 1, 1}
            }
        };

        runTests(testCasesF);
    }

    public void runTests(int[][][] testCasesF) {
        for(int[][] testCase : testCasesF) {
            test(testCase, testCaseSolutionOutput(testCase));
        }
    }

    private boolean test(int[][] f, String[] expected) {
        Solution sol = new Solution();
        String[] output = sol.shortestPath(f);
        boolean passed = Arrays.equals(output, expected);
        System.out.println("Forest: " + Arrays.deepToString(f));
        System.out.println("Expected: " + Arrays.toString(expected));
        System.out.println("Your Output: " + Arrays.toString(output));
        System.out.println(passed ? "Test PASSED" : "Test FAILED");
        System.out.println("-----------");
        return passed;
    }

    private String[] testCaseSolutionOutput(int[][] f) {
        int rows = f.length;
        int cols = f[0].length;
        int[] start = null;
        boolean[][] visited = new boolean[rows][cols];
        int[][] directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
        
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (f[i][j] == 3) {
                    start = new int[]{i, j};
                    break;
                }
            }
        }
        
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[2] - b[2]);
        pq.offer(new int[]{start[0], start[1], 0});
        
        while (!pq.isEmpty()) {
            int[] current = pq.poll();
            int row = current[0];
            int col = current[1];
            int dist = current[2];
            
            if (visited[row][col]) continue;
            visited[row][col] = true;
            
            if (f[row][col] == 2) {
                return new String[]{String.valueOf(row), String.valueOf(col)};
            }
            
            for (int[] dir : directions) {
                int newRow = row + dir[0];
                int newCol = col + dir[1];
                
                if (newRow >= 0 && newRow < rows && newCol >= 0 && newCol < cols 
                    && !visited[newRow][newCol] && f[newRow][newCol] != 1) {
                    
                    int newDist = dist;
                    if (f[newRow][newCol] == 4) {
                        newDist += 2;
                    } else {
                        newDist += 1;
                    }
                    
                    pq.offer(new int[]{newRow, newCol, newDist});
                }
            }
        }
        
        return new String[]{};
    }

}
